packages/@aws-cdk/aws-servicecatalog/README.md

```diff
@@ -184,6 +184,48 @@ const product = new servicecatalog.CloudFormationProduct(this, 'Product', {
 });
 ```
 
+We can also define a `versioningStrategy` for your `ProductStack` deployment. 
+With the `Default` strategy, a new `ProductStack` is needed for each `productVersion`
+and each `productVersion` will get overwritten with the latest changes to your `ProductStack`.
+With the `RetainPreviousVersions` strategy, a local template will be retained for each `ProductStack` deployed.
+Subsequently, any `ProductStack` changes will not overwrite the current version.
+A new `productVersionName` must be specified in order for `ProductStack` changes to be deployed.
+When the `ProductStack` is updated with a new version, the previous version can still be deployed.
+The template of previously deployed `ProductStack` can be referenced using `fromProductStackContext`
+and passing the corresponding `ProductStack` id.
+
+After using the `RetainPreviousVersions` strategy to deploy version `v1` of your `ProductStack`,
+the `ProductStack` is updated to version `v2` and version `v1` can be deployed from context:
+
+```ts
+import * as s3 from '@aws-cdk/aws-s3';
+import * as cdk from '@aws-cdk/core';
+
+class S3BucketProduct extends servicecatalog.ProductStack {
+  constructor(scope: cdk.Construct, id: string) {
+    super(scope, id);
+
+    new s3.Bucket(this, 'BucketProductV2');
+  }
+}
+
+const product = new servicecatalog.CloudFormationProduct(this, 'MyFirstProduct', {
+  productName: "My Product",
+  owner: "Product Owner",
+  productVersions: [
```

Comment:
Since it looks like the format of this ought to be:

* There can only be 1 `fromProductStack`. Its versioningStrategy can be either DEFAULT (while iterating) or RETAIN_PREVIOUS_VERSIONS (when done)
* The rest MUST be `fromProductStackContext` 

I think an abstraction on top might be useful, because this is easy to get wrong.

I wonder if its worth a new construct class, `VersionedCloudFormationProduct`, which hides all those machinations.

Reply:
```ts
new VersionedCloudFormationProduct(this, 'MyProduct', {
  productName: "My Product",
  owner: "Product Owner",
  currentVersionTemplate: servicecatalog.CloudFormationTemplate.fromProductStack(...),
  currentVersionMutable: true | false,
  currentVersionName: 'v3',
});
```

The construct will itself retain enough state on disk to reconstruct current and past versions from the files it stores.

Reply:
The originally idea was actually to have it fully automated and abstracted away. (not sure if its the same as you are proposing).
This idea was shot down because we don't want to forcibly retain or deploy anything for the customer.  
It can also lead to some unexpected errors with everything being in sync.
In the above example, are `v1` and `v2` automatically retained and deployed. 
The user should be able to explicitly choose what they want to deploy and it should be written in code.

Also there's no reason why there can only be 1 `fromProductStack`, there could be multiple, there could be combinations of `fromProductStack` and `fromAsset` or `fromUrl` and they would all fall under the same Product.

Reply:
> Also there's no reason why there can only be 1 `fromProductStack`, there could be multiple, there could be combinations of `fromProductStack` and `fromAsset` or `fromUrl` and they would all fall under the same Product.

There *could* be ... but is that what people are likely to do in practice? In practice, wouldn't customers have different Products, that are versioned differently?